tardis/opacities/macro_atom/macroatom_solver.py

```diff
@@ -0,0 +1,133 @@
+        """
+
+        self.initialize = initialize
+        self.normalize = normalize
+
+    def initialize_non_markov_transition_probabilities(self, atomic_data):
+        """initialize the transition probabilitiy coefficients and block references when solving the first time
+
+        Parameters
+        ----------
+        atomic_data : tardis.io.atom_data.AtomData
+            Atomic Data
+        """
+
+        coef_and_block_ref = initialize_non_markov_transition_probabilities(
+            atomic_data
+        )
+        self.transition_probability_coef = coef_and_block_ref[
+            "transition_probability_coef"
+        ]
+        self.block_references = coef_and_block_ref["block_references"]
+        self.initialize = False
+
+    def solve_non_markov_transition_probabilities(
+        self,
+        atomic_data,
+        legacy_plasma,
+        tau_sobolev,
+        stimulated_emission_factor,
+    ):
+        """Solve the basic transition probabilities for the macroatom
+
+        Parameters
+        ----------
+        atomic_data : tardis.io.atom_data.AtomData
+            Atomic Data
+        legacy_plasma : tarids.plasma.BasePlasma
+            legacy base plasma
+        tau_sobolev : pd.DataFrame
+            Expansion Optical Depths
+        stimulated_emission_factor : np.ndarray
+
+        Returns
+        -------
+        pd.DataFrame
+            Transition Probabilities
+        """
+        if self.initialize:
+            self.initialize_non_markov_transition_probabilities(atomic_data)
+
+        non_markov_transition_probabilities = (
+            calculate_non_markov_transition_probabilities(
+                atomic_data,
+                legacy_plasma.beta_sobolev,
+                legacy_plasma.j_blues,
+                stimulated_emission_factor,
+                tau_sobolev,
+                self.transition_probability_coef,
+                self.block_references,
+                normalize=self.normalize,
+            )
+        )
+
+        return non_markov_transition_probabilities
+
+    def solve(
+        self,
+        legacy_plasma,
+        atomic_data,
+        tau_sobolev,
+        stimulated_emission_factor,
+    ):
+        """Solved the Macro Atom State
+
+        Parameters
+        ----------
+        legacy_plasma : tarids.plasma.BasePlasma
+            legacy base plasma
+        atomic_data : tardis.io.atom_data.AtomData
+            Atomic Data
+        tau_sobolev : pd.DataFrame
+            Expansion Optical Depths
+        stimulated_emission_factor : pd.DataFrame
+
+        Returns
+        -------
+        tardis.opacities.macroatom_state.MacroAtomState
+            State of the macro atom ready to be placed into the OpacityState
+        """
+
+        transition_probabilities = (
+            self.solve_non_markov_transition_probabilities(
+                atomic_data,
+                legacy_plasma,
+                tau_sobolev,
```

Comment:
that is part of the opacity state - I presume?

Reply:
what do you mean?